src/cache.rs

```diff
@@ -673,65 +680,61 @@ impl Cache {
-                                transformed_term,
-                                resolved_ids,
-                                import_errors,
-                            } = import_resolution::tolerant::resolve_imports(term, self);
-                            (transformed_term, resolved_ids, import_errors)
+                            import_resolution::tolerant::resolve_imports(term, self)
                         }
                         ErrorTolerance::Strict => {
-                            let (term, pending) =
-                                import_resolution::strict::resolve_imports(term, self)?;
-                            (term, pending, Vec::new())
+                            import_resolution::strict::resolve_imports(term, self)?.into()
                         }
                     };
 
-                    self.terms.insert(
-                        file_id,
-                        CachedTerm {
-                            term,
-                            state: EntryState::ImportsResolving,
-                            parse_errs,
-                        },
-                    );
+                    // unwrap!(): we called `unwrap()` at the beggining of the enclosing if branch
+                    // on the result of `self.terms.get(&file_id)`. We only made recursive calls to
+                    // `resolve_imports` in between, which don't remove anything from `self.terms`.
+                    let cached_term = self.terms.get_mut(&file_id).unwrap();
+                    cached_term.term = transformed_term;
+                    cached_term.state = EntryState::ImportsResolving;
+
+                    let mut done = Vec::new();
+
+                    // Transitively resolve the imports, and accumulate the ids of the resolved
+                    // files along the way.
+                    for id in pending {
+                        if let CacheOp::Done((mut done_local, _)) = self.resolve_imports(id)? {
+                            done.push(id);
+                            done.append(&mut done_local)
+                        }
+                    }
 
-                    (
-                        pending
-                            .iter()
-                            .flat_map(|id| {
-                                if let Ok(CacheOp::Done((mut ps, _))) = self.resolve_imports(*id) {
-                                    ps.push(*id);
-                                    ps
-                                } else {
-                                    Vec::new()
-                                }
-                            })
-                            .collect(),
-                        errors,
-                    )
+                    self.update_state(file_id, EntryState::ImportsResolved);
+                    (done, import_errors)
                 } else {
-                    let pending = self.imports.get(&file_id).cloned().unwrap_or_default();
```

Comment:
This part didn't make a lot of sense to me. If we are in this branch, we are recursively resolving imports. Which means:

1. We shouldn't set the state to `ImportsResolved` yet at the end of the call
2. Resolving imports of entries inside `self.imports` doesn't make a lot of sense (those imports were resolved already, meaning they should have been treated already - or be on in pending queue. It seems like the code might have been copy pasted from e.g. the typechecking code, where it makes sense (the imports indicate transitive imports that are yet to be typechecked).